### ISSUE TITLE
Map matching GPS traces and road attributes

### ISSUE THREAD

**Issue by mkromer** [valhalla/valhalla#2855]
I want to match a trace of GPS coordinates to the OSM road network to obtain road attributes, e.g., speed limit and road class at the given GPS coordinates. I would have thought that this is possible by using a trace_attributes call, querying matched.edge_index, edge.speed_limit and edge.road_class attributes. 

When performing a request like this 

`curl -X POST   http://localhost:8002/trace_attributes   -H 'Content-Type: application/json'   -H 'Postman-Token: c4323b77-7016-4086-808e-551e912b064c'   -H 'cache-control: no-cache'   -d '{"shape":[{"lat": 1.407722, "lon": 103.80849, "time": 1537460873}, {"lat": 1.395893, "lon": 103.81542, "time": 1537460933}, {"lat": 1.393112, "lon": 103.82495, "time": 1537460973}, {"lat": 1.393462, "lon": 103.830005, "time": 1537460993}], "filters": {"attributes": ["edge.road_class", "edge.speed_limit", "matched.point", "matched.type", "matched.edge_index", "edge.id", "matched.point"], "action": "include"}, "costing": "auto", "shape_match": "map_snap"}'` 

I get the following response

```
{'matched_points': [
  {'edge_index': 0, 'type': 'matched', 'lat': 1.407718, 'lon': 103.80848},
  {'edge_index': 0, 'type': 'matched', 'lat': 1.395872, 'lon': 103.815405},
  {'edge_index': 0, 'type': 'matched', 'lat': 1.393103, 'lon': 103.824949},
  {'edge_index': 13, 'type': 'matched', 'lat': 1.393434, 'lon': 103.83001}],
 'edges': [
  {'id': 293232197648, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 292728881168, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 294339493904, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 293836177424, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 876508889104, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 876374671376, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 1670540967952, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 294171721744, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 876240453648, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 645083971600, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 645151080464, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 292661772304, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 646224822288, 'road_class': 'motorway', 'speed_limit': 90},
  {'id': 1632221806608, 'road_class': 'motorway', 'speed_limit': 90}],
 'alternate_paths': []}
```
At first glance there are all road attributes I need. However, I cannot make the connection between my GPS points and the corresponding road segment (=edge) since the edge_index refers either to the first or last edge but not to any of the edges in between (which it should for the given GPS coordinates).

Am I missing something here?

**Comment by amolskh** [comment]
Even I am facing exact same issue.. Would be great if someone can guide us on this one

**Comment by mandeepsandhu** [comment]
I believe the `edge_index` in the list of `matched_points` points to the edge the corresponding input coordinate is matched to. As for the other edges, they might be intersecting edges along the path. CMIIW @kevinkreiser

**Comment by amolskh** [comment]
> I believe the `edge_index` in the list of `matched_points` points to the edge the corresponding input coordinate is matched to. As for the other edges, they might be intersecting edges along the path. CMIIW @kevinkreiser

Yes , but those indexes done correspond to routable connected edges . even though snapped path / points seem to be correct

**Comment by amolskh** [comment]
@kevinkreiser could this issue because of this TODO comments to be implemented https://github.com/valhalla/valhalla/blob/master/src/tyr/trace_serializer.cc#L379

any thoughs ?

**Comment by mkromer** [comment]
I have found that by using the 3.0.9 release - instead of the latest version I was using before - my problem is solved. Haven't tried 3.1.0.

**Comment by kevinkreiser** [comment]
it sounds like a bug to me i just havent had time to look into it yet... Sorry for the long delay, there are actually a few bug reports like this at this point :frowning_face:

**Comment by Blobonat** [comment]
Any updates on this?
